### DISCUSSION TITLE
SciPy 2021 Sprint Topics

### DISCUSSION THREAD

**Discussion by j9ac9k** [pyqtgraph/pyqtgraph#1874] [General]
Hello everyone,

PyQtGraph will be partaking in the [SciPy 2021 Sprints](https://www.scipy2021.scipy.org/sprints)!  We haven't heard much in the way of details regarding timelines (@deppen8 👀 ) but I think at this point due to the wide array of time-zones maintainers/contributors span, we should plan on having some kind of general presence during the weekend of July 17/18.  

EDIT:

[Sprint schedule is up](https://www.scipy2021.scipy.org/schedule), sprints will start at Friday July 16 at 1800 CDT for the APAC time zone, and run until 1700 Sunday July 18 CDT.

There will be a 4-hour "mentored" time zone blocks, 1 block per different region, times look to be

* July 16 1800-2200 CDT for APAC
* July 17 0500-0900 CDT for European Time Zones
* July 17 1100-1500 CDT for American Time Zones

The rest of the weekend the Discord server that communication will be happening over will remain up.   I plan on being online contributing during the bulk of the weekend, and "mentoring" during the American time zone hour block.  If you're interested in "mentoring" new-comers to the project, please sign up [here](https://docs.google.com/forms/d/e/1FAIpQLScEW0TPZg378YymDr2NSN0Q9UgY-kFjsEDY5HFHDp3IB6FmHg/viewform).

Some other Relevant links: 

https://mentored-sprints.netlify.app/

End EDIT

Which begs the question, what should we work on?  From what I can tell, this will be the first time a majority of the maintainers will be collaborating on the library at the same time.

Some ideas:

1. Add sufficient pyqtgraph functionality to support being a plotting [backend for pandas](https://pandas.pydata.org/docs/dev/development/extending.html#plotting-backends)
2. Modify our (network) graph-plot code to better integrate with NetworkX
3. Expand numba usage throughout the library (downsample methods immediately come to mind!)
4. Work on a gallery (the example app is great for minimum working examples, but we should have something more to show-off the plotting capability, not just demonstrate it).
5. Something that has been nagging at me is consistent benchmarking/FPS measurement.  A lot of our benchmarks seem to use slightly different mechanisms to handle FPS determination, perhaps we should standardize that?
6. Expand the test suite for far greater coverage?
7. Introduce type-hinting
8. Address the PR backlog (at least some of the scarier PRs)
9. Document how to use with jupyter notebooks (good first issue perhaps?)

This list is just some of the things I have had on my mind.

I'd love to hear from others on this topic.  In the meantime, I should probably update the contributing guide, and get to identifying more issues as good first issues.

I'm going to tag some folks here as I'd love to get input from others on what we should work on during this time.

@outofculture @ixjlyons @NilsNemitz @ntjess @ksunden @pijyoi

**Comment by outofculture** [comment]
I'd love to work on adding more numba, and I'd like to see if the combination of numba and cupy is at all a promising line of performance expansion on ImageItem and its ilk. I'd also be willing to try to add cupy processing to more of the library, if we can identify likely fruitful areas where the cost of copying the data back and forth is worth it for the parallelization.

What other tasks are improved with lots of attention and breadth of experience? Applying automated styling? Type hints?

**Reply by j9ac9k** [reply]
I don't have a good feel for what kind of tasks would be good to use for cupy.  I assume it would largely effectively pure numpy methods involving large arrays and large degrees of vectorization?  Only other thing I can think that might benefit here from cupy support is the NaN checks/filtering in `arrayToQPath`.

We have a draft PR for automated styling where @NilsNemitz could likely use more feedback on.  

Type-hinting might be a good thing to discuss, if anything, to develop a consensus on how to do it effectively for a library of this size.

**Comment by ixjlyons** [comment]
Re the gallery: I'm particularly interested in getting this going. Right now the only blocker is figuring out what to do with our examples. Just enabling the gallery on our entire examples folder causes a few problems so I've been thinking about how we might split them up into categories. I'm also interested in turning at least some of them into tutorials. This _could_ be done in phases where we start by leaving the examples as-is and write new demos/tutorials for the gallery and then gradually migrate. This likely deserves its own discussion thread.

On the other hand, it doesn't necessarily need to be a focus of the sprint -- I can chip away at it any time. It could be nice to have a few people available real-time for decisions/questions though. In any case, I'm supportive of any topic(s) we land on and plan to contribute regardless.

**Reply by j9ac9k** [reply]
I thought you might have something to suggest here, I think doing more complicated visualizations, and extending them into tutorials is definitely a good thing to do.  On one hand, definitely not something we need a number of us to work on at the same time so perhaps not a great sprint topic (but perhaps we can agree what kind of visualizations we want to have in a gallery app and then move on?  I not-low-key still want to do my n-body orbiting simulation ...might want to start on that.

**Reply by ixjlyons** [reply]
I'll start a discussion with a few things I could use some feedback on and work on it async (deciding on a couple things is the blocker). I agree it's kind of an individual task, at least getting things set up, so it probably makes more sense to focus on a couple collective efforts for the sprint.

**Comment by ixjlyons** [comment]
Another potential topic: fully address the PR backlog. I think the hardest part about reviewing an older PR is absence of feedback from others who've looked at it in detail. This could be an interesting opportunity to get a couple people on each and knock them out. Doing it right after 0.12.2 and targeting 0.13.0 seems like it'd give some breathing room too since I always worry about unintended side effects of what seem like straightforward changes.

**Reply by j9ac9k** [reply]
> Another potential topic: fully address the PR backlog. I think the hardest part about reviewing an older PR is absence of feedback from others who've looked at it in detail. This could be an interesting opportunity to get a couple people on each and knock them out. Doing it right after 0.12.2 and targeting 0.13.0 seems like it'd give some breathing room too since I always worry about unintended side effects of what seem like straightforward changes.

this ....as boring as it sounds, might be a good thing to do...  More specifically, I think we should take the opportunity for many of us to look over @NilsNemitz 's #1625 it's a monster PR with a fantastic feature and @0000matteo0000 's #1359 as this is very frequently brought up.

both these features are very much in demand, the authors are motivated to get them merged, we've been shying away largely due to complexity IMO.

**Reply by ixjlyons** [reply]
Yep it's definitely not as fun as adding new stuff... Adding test coverage actually sounds more fun to me personally. Maybe tough as a main sprint topic for newer contributors though.

That is a good idea to focus on specific PRs -- it's probably only realistic to close out a few bigger ones or a handful of lower hanging fruit over a weekend really.

**Comment by j9ac9k** [comment]
Not specific to a topic, but if anyone wants to volunteer to be a mentor during the sprints, please considering [signing up here](https://docs.google.com/forms/d/e/1FAIpQLScEW0TPZg378YymDr2NSN0Q9UgY-kFjsEDY5HFHDp3IB6FmHg/viewform)

More information on the mentoring can be read [here](https://mentored-sprints.netlify.app/) (TL;DR you sign up to make yourself available to help newcomers to the project for a time-block specific to the mentored time-zone block.  See the original post for the times the blocks occur.

If I'm reading this right, sounds like some volunteers at SciPy will aim to provide general help desk support, but project specific mentors can provide support/help for onboarding people to a specific project, and more project specific issues....

**Comment by j9ac9k** [comment]
Just updated the original post with some links/times.

**Comment by outofculture** [comment]
[This doc](https://mentored-sprints.netlify.app/mentors/02-prepare/) recommends we do some issue tagging ahead of the sprints. I'll try to do some of that next week.

**Comment by j9ac9k** [comment]
And the sprint was a great success.  Seems like convergence was a combination of @pijyoi addressing a series of issues in GLViewWidget and everyone else doing their part to address the PR queue, and for a brief few hours, we tied our lowest number of open PRs in probably a decade 😆   

While disappointing that no new contributors joined us this sprint, I think this weekend was a fantastic opportunity for us to collaborate together in a way we haven't done historically, and we should try to do so again.  Perhaps we can try and make a strong pitch for newcomers to the project if we have a gallery we can highlight, not just the example app.

I think trying to set aside a weekend like this would work once every 3 months or so, and potentially try and time it with a sprint from a large conference (PyCon, SciPy, maybe some PyCascades?)  What does everyone else think?
